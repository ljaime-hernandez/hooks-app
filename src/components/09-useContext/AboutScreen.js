import React from 'react'

export const AboutScreen = () => {
    return (
        <div>
            <h1>About Screen</h1>
        </div>
    )
}
