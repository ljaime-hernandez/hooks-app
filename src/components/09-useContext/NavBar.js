import React from 'react';
import {Link, NavLink} from 'react-router-dom';

export const NavBar = () => {

    // the NavBar component was designed to have a user-friendly display, the NavLink component
    // is exported from the react-router-dom and it comes with a ternary operation in its className,
    // which will highlight the link component where the user is located, if the user is not located in the
    // respective component then the text on the NavBar text will appear slightly faded
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
  
          <Link className="navbar-brand" to="/">UseContext</Link>
          
          <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
            <div className="navbar-nav">
              <NavLink className={`nav-link ${ (navData) => navData.isActive && 'active' }`}  to="/">Home Screen</NavLink>
              <NavLink className={`nav-link ${ (navData) => navData.isActive && 'active' }`}  to="/login">Login Screen</NavLink>
              <NavLink className={`nav-link ${ (navData) => navData.isActive && 'active' }`}  to="/logout">Logout Screen</NavLink>
            </div>
          </div>

      </nav>
    )
}
