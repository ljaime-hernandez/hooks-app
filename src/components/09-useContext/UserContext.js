import {createContext} from 'react';

// higher-order component where the "Screen" components will
// exchange information about the user created in the MainApp, further
// explanation of the context can be found in each component
export const UserContext = createContext(null);