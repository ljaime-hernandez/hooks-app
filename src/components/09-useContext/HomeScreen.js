import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

// i use the useContext hook which utilize the custom UserContext designed for us to
// have a "intermediary file" which will handle information used by all our components
// removing the need of a parent-child component communication. Instead, this and other
// components will communicate with the context to retrieve the information required to
// render or validate additional information.
const {user} = useContext(UserContext);

    return (
        <div>
            <h1>Home Screen</h1>

            <pre>
                {/*just a small render of the user information created on a different component
                    which has direct communication with the context as well */}
                {JSON.stringify(user, null, 4)}
            </pre>
        </div>
    )
}
