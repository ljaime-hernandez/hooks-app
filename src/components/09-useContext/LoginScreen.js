import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {

    const {setUser} = useContext(UserContext);
    // the component below will use the destructured context value, which in this case
    // is a setState function for us to set a default user value, it will be rendered in
    // both the HomeScreen and the LogoutScreen components
    return (
        <div>
            <h1>Login Screen</h1>
            <button
                className='btn btn-primary'
                onClick={() => setUser({
                    id: new Date().getTime(),
                    name: 'Miguel'
                })}
            >
                Create user
            </button>
        </div>
    )
}
