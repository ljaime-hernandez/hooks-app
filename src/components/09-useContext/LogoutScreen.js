import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LogoutScreen = () => {

    const {user, setUser} = useContext(UserContext);
    
    // same as in the LoginScreen component, the context values are destructured
    // in this case i mix both the HomeScreen Rendering and the setUser function 
    // to both render the user information and change the user values to an empty
    // state, just as a small demonstration on how to "reset" the information
    // in the whole context 
    const handleClick = () => {
        setUser({});
    }

    return (
        <div>
            <h1>Logout Screen</h1>
            <pre>
                {JSON.stringify(user, null, 4)}
            </pre>
            <button
                className='btn btn-warning'
                onClick={handleClick}
            >
                Logout
            </button>
        </div>
    )
}
