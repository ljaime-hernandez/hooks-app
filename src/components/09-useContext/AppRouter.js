import React from "react";
import {
  BrowserRouter as Router,
  Route,
  Routes,
  Navigate
} from "react-router-dom";
import { LogoutScreen } from "./LogoutScreen";
import { HomeScreen } from "./HomeScreen";
import { LoginScreen } from "./LoginScreen";
import { NavBar } from "./NavBar";

// on this component i demonstrate how a Router component (exported through the installation of a 
// react-router-dom dependency) works and how it allow us to, in this case, do a small demo for a 
// SPA (Single Page Application).
// The NavBar is a custom nav component which will include links to our different components and
// additional features explained on it
// The Routes component is an export brought from the react-route-dom dependency, which will allow
// us to handle the different routes contained in this folder, separated on different components
// The Route component is also an export from react-route-dom, used solely to handle the path leading
// towards single components, each route will render an specific component described on its element
// attribute, but i also created an additional route (the last one) which will handle any additional
// route written by the user which does not lead to any of the components created on this context.
export const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar/>
                <Routes>
                    <Route  path="/" element={<HomeScreen/>}/>
                    <Route  path="/logout" element={<LogoutScreen/>}/>
                    <Route  path="/login" element={<LoginScreen/>}/>
                    <Route  path="*" element={<Navigate replace to="/" />} />    

                </Routes>
            </div>
        </Router>
    )
}
