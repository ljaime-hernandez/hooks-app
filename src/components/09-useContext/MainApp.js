import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

   const [user, setUser] = useState({});

   // the useState is created just so we have a main source from where to retrieve
   // the main user information, given as well with the setUser function,
   // which both will be passed as arguments to our custom UserContext component.
   // The UserContext then will be the middle component from which the rest of the "Screen" 
   // components will retrieve the main user information for them to render, create or 
   // reset the user information, according to its description.
   // As shown below, for that to work we have to encapsulate the main AppRouter component 
   // with the context, as the context will be the "father" component containing the main 
   // information as described above.
    return (
            <UserContext.Provider value={{
                user: user,
                setUser,
            }}>
                <AppRouter/>
            </UserContext.Provider>
    )
}
