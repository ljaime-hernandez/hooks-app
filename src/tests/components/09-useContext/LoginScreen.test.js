import { mount } from "enzyme";
import { UserContext } from "../../../components/09-useContext/UserContext";
import { LoginScreen } from "../../../components/09-useContext/LoginScreen";

/* to run this test:
1. run the 'npm install' command from the hooks-app folder 
2. run the 'npm install --save-dev enzyme' command (if you havent done so)
3. run the 'npm install --save-dev enzyme-to-json' command (if you havent done so)
4. run the 'npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 --legacy-peer-deps' command (if you are using React 17 as i do)
5. run the 'npm install --save-dev @testing-library/react-hooks' command (if you havent done so)
6. make sure the setupTests.js file include the enzyme, enzyme-to-json and the react adapter libraries
7. run the command 'npm run test'
8. to have a clearer view of this single js test file, press p. then type the file name 'LoginScreen.test.js'
*/

describe('Tests on LoginScreen component', () => {

    // the setUser is passed to the LoginScreen component as parameter for it to use
    // it on the user setup
    const setUser = jest.fn();

    // in the actual component, we pass both the user object and the setUser function,
    // in this component we dont render the user object but we just call the function to
    // assign the values to it.
    // We use the mount method from enzyme instead of the shallow, this is because the 
    // shallow method will render only the first component containing the rest of the
    // components, here it would only render the UserContext but it would not show any
    // information but the LoginScreen component tag, we need to render the LoginScreen 
    // with this procedure as it depends on the context itself, the mount will allow us
    // to fix this inconvenience
    const wrapper = mount(
            <UserContext.Provider 
                value={{setUser}}
            >
                <LoginScreen />
            </UserContext.Provider>)

    test('should render properly', () => {
        
        expect(wrapper).toMatchSnapshot();
    });

    test('should execute setUser with the appropriate argument', () => {
        
        // the setUser is just a callback function that assigns values to an object,
        // as seen in the actual component, the default values are set in the click event,
        // theres no input or submit event going on it, thats why we can put the parenthesis
        // syntax in front of this onClick property which will call what is written in the
        // original component.
        // The LoginScreen component will generate a number based on a Date and time, 
        // this will generate 'random' numbers which we dont really need to specify at the moment,
        // thats why the object returned by the setUser function can be evaluated in the id property
        // as any, the name is specific so we do test it accordingly
        wrapper.find('button').prop('onClick')();
        expect(setUser).toHaveBeenCalledWith({
            id: expect.any(Number),
            name: 'Miguel'
        });
    })
})