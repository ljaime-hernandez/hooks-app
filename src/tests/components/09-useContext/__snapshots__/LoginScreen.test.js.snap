// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginScreen component should render properly 1`] = `
<div>
  <h1>
    Login Screen
  </h1>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Create user
  </button>
</div>
`;
